app/validators/digital_object/enabled_dynamic_fields_validations.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+module DigitalObject::EnabledDynamicFieldsValidations
+  # Returns any errors related to project level validations.
+  #
+  # @return [String] if there are errors
+  # @return false if there are no errors
+  def enabled_field_errors(digital_object, data)
+    errors = []
+    dynamic_field_paths = []
+    data.each do |df_group, children|
+      dynamic_field_paths = collect_field_paths(df_group, children, dynamic_field_paths)
+      dynamic_field_paths.each do |dynamic_field_path|
+        if (e = not_enabled?(dynamic_field_path, digital_object))
+          errors.concat e.map { |i| [dynamic_field_path, i] }
+        end
+      end
+    end
+
+    errors = check_required_fields(dynamic_field_paths, digital_object, errors)
+
+    errors
+  end
+
+  def not_enabled?(dynamic_field_path, digital_object)
+    no_match = true
+    dynamic_field = DynamicField.find_by(path: dynamic_field_path)
+    if dynamic_field && EnabledDynamicField.where(dynamic_field: dynamic_field.id,
+                                project: digital_object.primary_project,
+                                digital_object_type: digital_object.digital_object_type).first
+      no_match = false
+    end
+    return ['field must be enabled'] if no_match
+    false
+  end
+
+  def check_required_fields(dynamic_field_paths, digital_object, errors)
+    required_enabled_fields = EnabledDynamicField.where(project: digital_object.primary_project,
+                                  digital_object_type: digital_object.digital_object_type, required: true)
+
+    required_enabled_fields.each do |enabled_field|
+      errors.concat ['is required'].map { |i| [enabled_field.dynamic_field.path, i] } unless dynamic_field_paths.include?(enabled_field.dynamic_field.path)
+    end
+
+    errors
+  end
```

Comment:
```suggestion
  # For the given project and digital_object type, checks the given present_dynamic_field_paths list
  # to see if it contains all of the required fields.
  #
  # @param dynamic_field_paths [Array<String>] An array of dynamic field paths to check.
  # @param project [Array<String>] An array of dynamic field paths to check.
  # @return [Array<Array<String>>] An array errors of the format: [['field name', 'error message], ['field name', 'error message]]
  def check_missing_required_fields(project, digital_object_type, present_dynamic_field_paths)
    required_field_paths = DynamicField.where(
      id: EnabledDynamicField.where(
        project: project, digital_object_type: digital_object_type, required: true
      ).pluck(:dynamic_field_id)
    ).pluck(:path)

    [required_field_paths - present_dynamic_field_paths].each do |missing_required_field_path|
      [missing_required_field_path, 'is required']
    end
  end
```
In an earlier comment, I mentioned how this method modifies errors and also returns it.  Here's a proposed new version of the method that only returns a set of errors, and also gives you a more efficient record selection method (that avoids extra queries that I think would have been run each each iteration for `enabled_field.dynamic_field`).  I changed the method signature because after removing `errors`, I realized that the method didn't actually require an entire digital_object instance to work -- it only needed a reference to a project and digital_object_type string.  That should make it easier to test, and less coupled to a more complex object.

Reply:
Before making this suggested change, definitely check out my other comments because what I proposed here may not be exactly what you want.